Sources/Signal.swift

```diff
@@ -150,13 +150,21 @@ public final class Signal<Value, Error: Swift.Error> {
 				if case let .alive(state) = signal.status {
 					state.send(event)
 
+					// Check if the status has been bumped to `terminating` due to a
+					// concurrent or a recursive termination event.
+					//
+					// Note that as the check happens before acquiring `statusLock`, it
+					// could be a false positive, and thus a second check after the lock
+					// is acquired is mandatory.
 					if case .terminating = signal.status {
-						// Acquire `statusLock`, and handle the termination if the signal
-						// is still at `terminating` state.
+						// Acquire `statusLock`. If the termination has still not yet been
+						// handled, take it over and bump the status to `terminated`.
 						signal.statusLock.lock()
+
 						if case let .terminating(state) = signal.status {
```

Comment:
The double checking is to free the normal event delivery from having to acquire `statusLock`, by exploiting wait-free reading of read-copy-update.

Concurrent senders perform a relaxed read from `status`, and enter the termination handling routine only if they see `terminating`. However, as this is relaxed and some other may have already handled it and bumped the status to `terminated`, a second check is performed after acquiring `statusLock`.